x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -519,44 +510,43 @@ private void authorizeBulkItems(RequestInfo requestInfo, AuthorizationInfo authz
         final Map<String, Set<String>> actionToIndicesMap = new HashMap<>();
         final AuditTrail auditTrail = auditTrailService.get();
 
-        authorizedIndicesSupplier.getAsync(ActionListener.wrap(authorizedIndices -> {
-            resolvedIndicesAsyncSupplier.getAsync(ActionListener.wrap(overallResolvedIndices -> {
-                final Set<String> localIndices = new HashSet<>(overallResolvedIndices.getLocal());
-                for (BulkItemRequest item : request.items()) {
-                    final String itemAction = getAction(item);
-                    String resolvedIndex = resolvedIndexNames.computeIfAbsent(item.index(), key -> {
-                        final ResolvedIndices resolvedIndices =
-                            indicesAndAliasesResolver.resolveIndicesAndAliases(itemAction, item.request(), metadata, authorizedIndices);
-                        if (resolvedIndices.getRemote().size() != 0) {
-                            throw illegalArgument("Bulk item should not write to remote indices, but request writes to "
+        resolvedIndicesAsyncSupplier.getAsync(ActionListener.wrap(overallResolvedIndices -> {
+            final Set<String> localIndices = new HashSet<>(overallResolvedIndices.getLocal());
+            for (BulkItemRequest item : request.items()) {
+                final String itemAction = getAction(item);
+                String resolvedIndex = resolvedIndexNames.computeIfAbsent(item.index(), key -> {
+                    final ResolvedIndices resolvedIndices =
+                            indicesAndAliasesResolver.resolveIndicesAndAliases(itemAction, item.request(), metadata, localIndices);
```

Comment:
The change here relies on the fact that the last argument `authorizedIndices` is not used in `resolveIndicesAndAliases`. Therefore, we can pass `localIndices` to it. In fact, we can even just pass `null`. I think this effect is somewhat hidden and hard for another person to comprehend. It works for now. But I am concerned that it could get misused in future (e.g. due to refactoring etc). Can we make it more obvious? I don't have great suggestions, maybe have a separate method `IndicesAndAliasesResolver#resolveWritableIndices`?

Also, the first sentence of the PR description
> This PR reuses the resolved indices of bulk shard requests for authorizing their items.

 is not entirely accurate because there is no actual reuse, but rather not needed at all, i.e. the last argument can be `null`. It would be great if it can be made clearer as well. Thanks!

Reply:
Yang, from this comment, I get the impression that you don't like the approach I took in this PR, but you have still approved.
My goal is to avoid passing the authorized indices list computed for the parent `BulkShardRequest` when resolving the indices for the containing `BulkItemRequest`s. As you say, what we replace it with is not important since it is not used.

I think we have two options:
* have a new method for `IndicesAndAliasesResolver#resolveIndicesAndAliases` for `BulItemRequests` only
* use the existing method but pass a different argument

These are the two big options that I have weighted.
I stand by my choice and I can get into details if you are interested, but I think it would help if you also state your preference so that we can compare them.

> > This PR reuses the resolved indices of bulk shard requests for authorizing their items.

> is not entirely accurate because there is no actual reuse, but rather not needed at all, i.e. the last argument can be null. It would be great if it can be made clearer as well. Thanks!

I think it is accurate because this is what the PR does, it reuses the said variable as an argument.
The fact that the parameter is not used and can be null is not a change that this PR does.

Reply:
To clarify myself: I am happy with the overall goal of this PR, i.e. IIUC, removing the need of having a separate `authorizedIndicesSupplier` and passing it around. That's why I ticked approval.

The concern that I have is only with what we should pass to `resolveIndicesAndAliases` now that `authorizedIndices` is no longer available in `authorizeBulkItems`. I think passing `localIndices` is not the right answer because:
1. The parameter is called `authorizedIndices` and this causes confusion
2. It works because the passed-in argument is never used for this specific case. I am aware that it is an existing behaviour that the method `resolveIndicesAndAliases` sometimes does not use `authorizedIndices`. But the difference is: before this change, the callers do not need to know whether the argument is going to be used by the callee. They work off the method signature (a contract) and just pass in whatever is asked consistently in all call sites. But the current change makes the caller aware of this implementation detail. If the method implementation changes in future, it could cause subtle bugs. I think if the caller relies on the fact that the argument is not used, we should make it clearer, e.g. passing a `null` instead of `localIndices`.
3. If we just read the changes literally, resolved indices indeed get "reused" as the method argument. But this "reuse" is superficial since underlying it is never used. It is not really much different than a placeholder to satisfy the method signature (and compiler). So if it is indeed more of a placeholder, we should make the point clearer.

Based on the above, I think passing `null` as the last argument to `resolveIndicesAndAliases` feels better. Imagine if we need to write javadocs for `resolveIndicesAndAliases`, a nullable argument is likely easier to explain than `localIndices`.

In longer term, we might still want to think about refactoring existing methods to help reduce the complexity and increases readibility. But I can understand that more refactoring is outside scope of this PR.